staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,521 @@
+// eg: if desiredNumQueues is increased, SetConfiguration reconciles by
+// adding more queues.
+func (fqs *FQScheduler) SetConfiguration(concurrencyLimit, desiredNumQueues, queueLengthLimit int, requestWaitLimit time.Duration) {
+	// TODO(aaron-prindle) verify updating queues makes sense here vs elsewhere
+
+	// lock required as method can change Queues which has its indexes and length used
+	// concurrently
+	fqs.Lock.Lock()
+	defer fqs.Lock.Unlock()
+
+	// Adding queues is the only thing that requires immediate action
+	// Removing queues is handled by omitting indexes >desiredNumQueues from
+	// chooseQueueIdx
+
+	numQueues := len(fqs.Queues)
+	if desiredNumQueues > numQueues {
+		fqs.addQueues(desiredNumQueues - numQueues)
+	}
+
+	fqs.concurrencyLimit = concurrencyLimit
+	fqs.desiredNumQueues = desiredNumQueues
+	fqs.queueLengthLimit = queueLengthLimit
+	fqs.requestWaitLimit = requestWaitLimit
+}
+
+// TimeoutOldRequestsAndRejectOrEnqueue encapsulates the lock sharing logic required
+// to validated and enqueue a request for the FQScheduler/FairQueueingSystem:
+// 1) Start with shuffle sharding, to pick a queue.
+// 2) Reject old requests that have been waiting too long
+// 3) Reject current request if there is not enough concurrency shares and
+// we are at max queue length
+// 4) If not rejected, create a packet and enqueue
+// returns true on a successful enqueue
+// returns false in the case that there is no available concurrency or
+// the queuelengthlimit has been reached
+func (fqs *FQScheduler) TimeoutOldRequestsAndRejectOrEnqueue(hashValue uint64, handSize int32) FQPacket {
+	// TODO(aaron-prindle) removing locking now and doing it all in Wait()
+	// fqs.Lock.Lock()
+	// defer fqs.Lock.Unlock()
+
+	//	Start with the shuffle sharding, to pick a queue.
+	queueIdx := fqs.ChooseQueueIdx(hashValue, int(handSize))
+	queue := fqs.Queues[queueIdx]
+	// The next step is the logic to reject requests that have been waiting too long
+	fqs.removeTimedOutPacketsFromQueue(queue)
+	// NOTE: currently timeout is only checked for each new request.  This means that there can be
+	// requests that are in the queue longer than the timeout if there are no new requests
+	// We think this is a fine tradeoff
+
+	// Create a packet and enqueue
+	pkt := &Packet{
+		DequeueChannel: make(chan bool, 1),
+		EnqueueTime:    fqs.clock.Now(),
+		Queue:          queue,
+	}
+	if ok := fqs.rejectOrEnqueue(pkt); !ok {
+		return nil
+	}
+	return pkt
+
+}
+
+// removeTimedOutPacketsFromQueue rejects old requests that have been enqueued
+// past the requestWaitLimit
+func (fqs *FQScheduler) removeTimedOutPacketsFromQueue(queue FQQueue) {
+	timeoutIdx := -1
+	now := fqs.clock.Now()
+	pkts := queue.GetPackets()
+	// pkts are sorted oldest -> newest
+	// can short circuit loop (break) if oldest packets are not timing out
+	// as newer packets also will not have timed out
+
+	// now - requestWaitLimit = waitLimit
+	waitLimit := now.Add(-fqs.requestWaitLimit)
```

Comment:
The quantity calculated here is not a limit on waiting; it is a limit on arrival time.